### PR TITLE
Iox #131 Remove std::function in expected and optional

### PR DESCRIPTION
**Pull request by mossmaurice**
Thanks @prasannabhat for the help with the service discovery

### CODE REVIEW

File: iceoryx_utils/include/iceoryx_utils/posix_wrapper/timer.hpp [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <condition_variable>
 #include <cstdint>
 #include <ctime>
+#include <functional>
```
Comment (elfenpiff) [inline iceoryx_utils/include/iceoryx_utils/posix_wrapper/timer.hpp:28]: Why did you add the functional header?

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 2]
```diff
@@ -60,6 +61,17 @@ ComplexType returnComplexType(ComplexType foo)
     return foo;
 }
 
+int SameSignature(function_ref<void(void)> callback)
+{
+    callback();
+    return -1;
+}
+
+int SameSignature(function_ref<int(int)> callback)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:70]: I know it is a bit pedantic but could you please also add unit tests for

- ```int SameSignature(function_ref<int(void)> callback)```
- ```int SameSignature(function_ref<void(int)> callback)```
- ```int SameSignature(function_ref<int(int, int)> callback)```

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp [hunk 3]
```diff
@@ -274,3 +286,10 @@ TEST_F(function_refTest, StoreInStdFunctionResultEqual)
     std::function<int()> sut(moep);
     EXPECT_THAT(sut(), Eq(37));
 }
+
+
+TEST_F(function_refTest, CallOverloadedFunctionResultsInCallOfCorrect)
```
Comment (elfenpiff) [inline iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:291]: Could you please also add a test for the void(void) function and check if the value is equal to ```-1```

File: iceoryx_utils/include/iceoryx_utils/posix_wrapper/timer.hpp (hunk 1 above)
Comment (mossmaurice) [reply iceoryx_utils/include/iceoryx_utils/posix_wrapper/timer.hpp:28]: Used to be implicitly included via `optional` before.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:70]: Added. However, `int SameSignature(function_ref<void(int)> callback)` would be ambiguous to `int SameSignature(function_ref<int(int)> callback)` as the return value is not part of the signature, so this one is left out.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp:291]: Replaced `void(void)` with `int(void)` to have similar test cases.

File: iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp]: You are right.
